js-miniapp-sdk/src/miniapp-error-type.ts

```diff
@@ -0,0 +1,70 @@
+/**
+ * This class is a representation of an error sent from MiniApp mobile SDK
+ */
+export class MiniAppError extends Error {
+  /**
+   * the MiniAppErrorType defining this error.
+   * If no type con be extracted from the input then the MiniAppError will be considered of type `Other`
+   */
+  type: MiniAppErrorType;
+
+  /**
+   * the raw input received from the SDK used to construct the MiniAppError
+   */
+  raw: string;
+
+  /**
+   * Takes the error input string sent from mobile SDK.
+   * Should be formatted as follow: "error_key: error message"
+   *
+   * @param errorInput error input string sent from mobile SDK
+   */
+  constructor(public errorInput: string) {
+    super();
+    this.raw = errorInput;
+    const messageArray = errorInput.split(': ');
+    const error: MiniAppErrorType =
+      MiniAppErrorType[messageArray[0] as keyof typeof MiniAppErrorType];
+    let errorMessage: string | undefined;
+    if (error) {
+      errorMessage = errorTypesDescriptions.get(error);
+      this.type = error;
+    } else {
+      this.type = MiniAppErrorType.Other;
+    }
+    if (errorMessage) {
+      this.name = error;
+      this.message = errorMessage;
+    } else {
+      this.name = error
+        ? error
+        : messageArray.length > 1
+        ? messageArray[0]
+        : MiniAppErrorType.Other;
+      messageArray.splice(0, 1);
+      this.message =
+        messageArray.length > 0 ? messageArray.join(': ') : errorInput;
+    }
+  }
+}
+
+/**
+ * Enum for supported SDK error types
+ */
+export enum MiniAppErrorType {
+  AudienceNotSupportedError = 'AudienceNotSupportedError',
```

Comment:
I was wondering if it would make more sense to make each type a custom error class? I saw something like this in the Firebase JS SDK:
`FirebaseError` (base error class): https://github.com/firebase/firebase-js-sdk/blob/5ad7ff2ae955c297556223e6cb3ad9d4b897f664/packages/util/src/errors.ts#L74
`FirebaseStorageError` (example error sub-class): https://github.com/firebase/firebase-js-sdk/blob/919413cd06478365e306476c8eb2ad2aa1f2d3ee/packages/storage/src/implementation/error.ts#L24

The advantage of this is that users can check the error type using `error instanceof FirebaseStorageError` and it also allows each error type to have custom properties.

So we could try something similar:
```
export class MiniAppErrorType extends Error {
  constructor(message: string) {
    super(message)
    this.name = "MiniAppErrorType"
    Object.setPrototypeOf(this, MyCustomError.prototype)
  }
}

export class AudienceNotSupportedError extends MiniAppErrorType {
  constructor() {
    super("The value passed for 'audience' is not supported.")
    this.name = "AudienceNotSupportedError"
    Object.setPrototypeOf(this, AudienceNotSupportedError.prototype)
  }
}
```

What do you think?

Reply:
I agree, this seems to be the standard way, I'll make an attempt with this in mind